daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -1099,31 +1098,76 @@ private[lf] object Speedy {
             }
           case TTyCon(tyCon) =>
             value match {
-              case V.ValueRecord(_, fields) =>
-                val lookupResult =
-                  assertRight(compiledPackages.pkgInterface.lookupDataRecord(tyCon))
+              case V.ValueRecord(_, sourceElements) => {
+                val lookupResult = assertRight(
+                  compiledPackages.pkgInterface.lookupDataRecord(tyCon)
+                )
+                val targetFieldsAndTypes: ImmArray[(Name, Type)] = lookupResult.dataRecord.fields
                 lazy val subst = lookupResult.subst(argTypes)
-                // TODO: https://github.com/digital-asset/daml/issues/17082
-                // - disallow dropping extra fields which have value other than None.
-                val actualValues = (lookupResult.dataRecord.fields.iterator zip fields.iterator)
-                  .map { case ((_, fieldType), (_, fieldValue)) =>
-                    go(AstUtil.substitute(fieldType, subst), fieldValue)
+
+                // This code implements the compatibility transformation used for up/down-grading
+                // And handles the cases:
+                // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
+                val numS: Int = sourceElements.length
+                val numT: Int = targetFieldsAndTypes.length
+
+                // traverse the sourceElements, "get"ing the corresponding target type
```

Comment:
Nit pick: maybe `"getting"` might be better here?

Reply:
It was deliberate. Refers to the `get` 4 lines down.